lib/compiler/src/beam_types.erl

```diff
@@ -347,8 +347,13 @@ subtract(#t_integer{elements={Min, Max}}, #t_integer{elements={N,N}}) ->
     end;
 subtract(number, #t_float{elements=any}) -> #t_integer{};
 subtract(number, #t_integer{elements=any}) -> #t_float{};
-subtract(list, cons) -> nil;
-subtract(list, nil) -> cons;
+subtract(#t_list{type=TypeA}=T, #t_cons{head=TypeB}) ->
+    case meet(TypeA, TypeB) of
+        TypeA -> nil;
```

Comment:
This seems to be unsafe for a non-proper list. If the list is not known to be proper, you can't be sure that the tail of the final cons is nil.

Reply:
The subtraction is on the list itself, so returning `nil` is correct even when the list isn't proper.

Thanks for pointing this out though, we should only do it when the cons cell is at least as wide as the one in the list (lists being `#t_cons{} | nil`), which includes whether they're proper.